### PR TITLE
[Dataset] Adopt new datagetter to databuffer

### PR DESCRIPTION
**Pull request by zhoonit**
- [Dataset] Adopt new datagetter to databuffer

```
**Changes Propsed in this PR**
1. Adopt new data getter to neuralnet / factory
2. Fix: bugs found in raw_file_data_producer
3. Fix: databuffer not feeding last iteration when error

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/dataset/databuffer.cpp [hunk 1]
```diff
@@ -251,11 +250,16 @@ DataBuffer::startFetchWorker(const std::vector<TensorDim> &input_dims,
...
     while (true) {
-      auto iteration = generator();
-      auto last = std::get<0>(iteration);
-      bq->wait_and_push(std::move(iteration));
-      if (last == true) {
-        break;
+      try {
+        auto iteration = generator();
+        auto last = std::get<0>(iteration);
+        bq->wait_and_push(std::move(iteration));
```
Comment (kparichay) [inline nntrainer/dataset/databuffer.cpp:256]: The validation checks for the data and its shape will be performed here? as the iteration is coming from external functions. If yes, please add TODO here or at appropriate location.
These checks will probably be removed when the data is pre-allocated.

File: nntrainer/dataset/databuffer_factory.cpp [hunk 2]
```diff
@@ -40,33 +46,42 @@ std::unique_ptr<DataBuffer> createDataBuffer(DatasetType type,
...
 std::unique_ptr<DataBuffer> createDataBuffer(DatasetType type, datagen_cb cb,
                                              void *user_data) {
+
+  std::unique_ptr<DataProducer> dp;
   switch (type) {
   case DatasetType::GENERATOR:
-    return std::make_unique<DataBufferFromCallback>(cb, user_data);
+    dp = std::make_unique<FuncDataProducer>(cb, user_data);
+    break;
   case DatasetType::UNKNOWN:
```
Comment (kparichay) [inline nntrainer/dataset/databuffer_factory.cpp:76]: For a later PR, check if this enum is needed anymore.

File: nntrainer/models/neuralnet.cpp [hunk 3]
```diff
@@ -660,43 +663,37 @@ int NeuralNetwork::train_run() {
...
-          throw;
-        }
-        std::cout << "#" << epoch_idx << "/" << epochs;
-        float loss = getLoss();
-        train_buffer->displayProgress(count++, loss);
-        training.loss += loss;
-      } else {
-        train_buffer->clear();
+      auto [last, ins, labels] = *train_buffer->fetch();
+      if (last) {
```
Comment (kparichay) [inline nntrainer/models/neuralnet.cpp:681]: !last?

File: test/ccapi/unittest_ccapi.cpp [hunk 4]
```diff
@@ -205,8 +205,8 @@ TEST(nntrainer_ccapi, train_with_config_01_p) {
   EXPECT_EQ(model->initialize(), ML_ERROR_NONE);
   EXPECT_NO_THROW(model->train());
 
-  EXPECT_NEAR(model->getTrainingLoss(), 4.434051, tolerance);
-  EXPECT_NEAR(model->getValidationLoss(), 2.910938, tolerance);
+  EXPECT_NEAR(model->getTrainingLoss(), 4.1389656, tolerance);
```
Comment (kparichay) [inline test/ccapi/unittest_ccapi.cpp:208]: i wonder why the accuracies changed.

File: nntrainer/dataset/databuffer.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/dataset/databuffer.cpp:256]: I am currently working on the pre-allocate scenario, so let me add a comment here instead.

File: nntrainer/dataset/databuffer_factory.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/dataset/databuffer_factory.cpp:76]: I guess we can't delete this as it's already part of capi.

File: test/ccapi/unittest_ccapi.cpp (hunk 4 above)
Comment (zhoonit) [reply test/ccapi/unittest_ccapi.cpp:208]: It's been a while but afai remember, it was because the shuffling has changed from handcrafted to std::shuffle.

File: nntrainer/dataset/databuffer_factory.cpp [hunk 5]
```diff
@@ -12,24 +12,30 @@
...
  * @brief Factory creator with constructor
  */
 std::unique_ptr<DataBuffer> createDataBuffer(DatasetType type) {
+
+  std::unique_ptr<DataProducer> dp;
   switch (type) {
   case DatasetType::FILE:
-    return std::make_unique<DataBufferFromDataFile>();
+    dp = std::make_unique<RawFileDataProducer>();
+    break;
```
Comment (jijoongmoon) [inline nntrainer/dataset/databuffer_factory.cpp:31]: we just create DatasetType::FILE without file path?

File: nntrainer/dataset/databuffer_factory.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/dataset/databuffer_factory.cpp:76]: I cant find it in the capi. Can you point me where?

File: test/ccapi/unittest_ccapi.cpp (hunk 4 above)
Comment (kparichay) [reply test/ccapi/unittest_ccapi.cpp:208]: Can you once turn off shuffle in the old and new code and verify that the values match for the tests where you modified the values. This is critical for correct working.

File: nntrainer/dataset/databuffer_factory.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/dataset/databuffer_factory.cpp:76]: My bad, we don't have it. As other enums does not have this, I thought we have this. I'll put this in #303

File: nntrainer/dataset/databuffer_factory.cpp (hunk 5 above)
Comment (zhoonit) [reply nntrainer/dataset/databuffer_factory.cpp:31]: Yes, for the file, there are two way of creating it.

1. without file path, and later set as a property
2. with file path, (can be overriden with a property)

File: test/ccapi/unittest_ccapi.cpp (hunk 4 above)
Comment (zhoonit) [reply test/ccapi/unittest_ccapi.cpp:208]: Confirming it's exactly correct with shuffling is off.

leaving a record to reproduce

TC name: nntrainer_capi_nnmodel.train_with_file_01_p

1. 2.3630399703979492 (train loss)
2. 2.2508399486541748 (val loss)
3. 8.3332996368408203 (val acc)

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM
